### ISSUE TITLE
External server and relative URLs in Beta 12

### ISSUE THREAD

**Issue by marcaube** [bdkjones/CodeKit2#172]
I have a problem with CodeKit serving static files (JS, CSS) with paths generated by an external server. If the web root of CodeKit server is not the same as `localhost`, the files are not found.

---

I have a Symfony2 project at `http://localhost/myapp/web/` and I configured CodeKit like this:

![screen shot 2014-01-18 at 11 48 44 am](https://f.cloud.github.com/assets/1478986/1947656/bd7da9c8-8060-11e3-970a-7da3cc57aac9.png)

Symfony2 generates paths to static files like this

``` html
<link rel="stylesheet" href="/myapp/web/css/styles.css" />
```

It seems like it generates the path relative to `http://localhost/` and not `http://mycomputer.local:5757/`.

With CodeKit server, `/myapp/web/css/styles.css` is served as `http://mycomputer.local:5757/myapp/web/css/styles.css`, which returns a 404, because it would be like making a request for `http://localhost/myapp/web/myapp/web/css/styles.css`

**Comment by bdkjones** [comment]
You've got your settings misconfigured. 

Given that your link element has this address: `href="/myapp/web/css/styles.css"`, then your server's document root CANNOT be `/myapp/web/`. Rather, your server's root is `http://localhost/`, which is what you need to enter in the External Server Address field.

If your server is truly set to serve `myapp/web/` as its document root, then the `href` property in your link tag is incorrect and should be:  `href="/css/styles.css"`

**Comment by marcaube** [comment]
I have to type `http://mycomputer.local:5757/myapp/web/` to get to my project? What's the use of the CodeKit server then?

We absolutely have to configure vhosts?

**Comment by bdkjones** [comment]
I need to see the project and your file structure. Please zip the project and post a download link here so I can take a look.

**Comment by bdkjones** [comment]
If I understand correctly, you have a situation like this going on.
1. Your website exists on disk at some path. Let that path be:
   
   /Users/marcaube/Desktop/site/
2. In the folder above, you've got content for your website project. The main index.html file for your website exists at this path:
   
   /Users/marcaube/Desktop/site/myapp/web/index.html
3. Your External Server is set to host the folder:
   
   /Users/marcaube/Desktop/site/
4. So to preview your website in your browser, using the server you've got set up, you go to:
   
   http://localhost/myapp/web
5. You go to that address because the index.html file for the website is NOT in the document root directory (the one that your server is set to host). 
6. If you enter the following address in your browser, you don't get your website, but rather a list of files and folders in the directory `/Users/marcaube/Desktop/site/`:
   
   http://localhost

Does the above scenario sound accurate?

**Comment by bdkjones** [comment]
IF the above description sounds accurate, then here's the deal:

The external server address field in CodeKit should be: `http://localhost`

That's the address to the webroot of your server. The fact that your index.html is in a subfolder means that, yes, you'll have to add /myapp/web to the end of the URL that CodeKit automatically opens in your browser when you click the Preview button. (CodeKit always opens your webroot folder when you click preview).

What you're missing is an .htaccess file or similar to redirect HTTP requests that target your webserver's root folder (http://localhost) to the subpath where you have your index file: (http://localhost/myapp/web).

**Comment by marcaube** [comment]
I couldn't attach the zip file because > 5mb but the structure is basically this [https://github.com/symfony/symfony-standard]().

Your scenario was pretty accurate yes, let's make a diagram...

```
~/www   - This is my External Server root folder, http://localhost/ points to here
    ├─ site 1/
    ├─ site 2/
    └─ myapp/
        ├─ app/       - My application configs are there, database, mail, logs and cache
        ├─ src/       - My modules (bundles) are there, that's the core of my application
        ├─ vendor/    - The vendor modules, like API clients and database ORM
        └─ web/       - The web root, where my front-controller is
            ├─ .htaccess
            ├─ index.php
            ├─ apple-touch-icon.png
            ├─ css/
            ├─ favicon.ico
            ├─ js/
            └─ robots.txt
```

The CSS and JS files are generally in the modules (`src/` or `vendor/`) and sym-linked to the `web/` folder.

---

> What you're missing is an .htaccess file or similar to redirect HTTP requests that target your webserver's root folder (http://localhost) to the subpath where you have your index file: (http://localhost/myapp/web).

Well, the problem is that I have multiple such projects in the root folder. With an `.htaccess` file in the root folder, I'd have to edit it every time I switch between projects.

Seems like the best for me is still to define vhosts for each of these projects and just don't use the CodeKit server and its magical auto-refresh.

**Comment by bdkjones** [comment]
Okay. When Symphony generates root-relative links for various assets on your page, it should generate them STARTING from the `myapp` folder, because that's the root of your project. So, the absolute URL to your stylesheet ought to be:

```
href="/web/css/styles.css"
```

That's the appropriate thing to do, because each of those projects in the `www` folder is a self-contained, separate website that should be hosted as a unit. 

Instead, Symphony is generating the root-relative links starting at whatever folder your server happens to currently be hosting. (In this case, `www`). That's backwards; the root-relative links should start at the project's root folder (`myapp`) and then you should configure your server to treat that folder as the webroot for hosting that project. (You can do that through vhosts, etc.)

Other platforms like Wordpress offer a configuration option to set `siteURL`, which tells the platform which folder forms the document root for the website. I suppose Symphony offers something similar, but I've never worked with it. 

### The bottom line

CodeKit expects the address you provide in "External Server Address" to be the document root for the project. So when it sees a root-relative link, it's going to find that file by taking your External Server Address, exactly as you specified it, and appending the path from the link in question. That's exactly what you're observing: 

You entered this external server address:

```
http://localhost/myapp/web
```

And CodeKit got a request for the following root-relative asset:

```
/myapp/web/css/styles.css
```

So it combined the two and looked for the file at this path:

```
http://localhost/myapp/web/myapp/web/css/styles.css
```

Because it treated your full external server address as the document root for the project.

**Comment by bdkjones** [comment]
That said, I don't want to deal with 5,782 emails regarding this once I launch the app, so I will add some advanced configuration options for situations like this. It's going to be a ton of work.

**Comment by marcaube** [comment]
Well, instead of working your ass off on an edge case, you could document the issue and give simple solutions. From what I understand, CodeKit use the proper folder but Symfony somehow "thinks" it's served from the apache server root. So maybe a little note somewhere about this issue when you release.

From the security standpoint, exposing a single php file from the `web/` directory reduces the attack surface...it's kinda wise if you ask me. A lot of projects (*cough* wordpress *cough*) can leave a lot of files/folders accessible from the web.

**So, possible solutions...**
1. I can create a vhost – like I was doing before – and point CodeKit to this address. But that's a vhost pointing to another vhost, so I'll borrow your words and say it's a little backwards...but it works and I get to keep the compile-refresh-magic of CodeKit.
2. There may also be a piece of .htaccess-fu that would do the trick but I have yet to find the right rules. I'll have to check how Wordpress works in that regard and read about RewriteBase, maybe I can craft a rule only for static files and only for the CodeKit url.
3. There is a framework config that could do the trick for most, but some less conventional part of my project still didn't quite work.

``` yml
# app/config/config.yml

framework:
    templating:
        assets_base_urls: http://localhost/myapp/web/
```

**Comment by bdkjones** [comment]
This is resolved in Beta 13.

When you get that beta, you'll want to enter "http://localhost" for your External Server Address and then, in the new textfield you'll see just below it, called "preview subpath", you'll enter: "myapp/web". 

When you click the preview button, the address that opens will be: http://[yourComputerName].local:5757/myapp/web 

But when CodeKit goes to resolve the root-relative URLs, it will resolve them from the External Server Address, meaning it will just work.
